modules/RTC/TrackStreamingStatus.js

```diff
@@ -0,0 +1,738 @@
+            this._clearRestoringTimer();
+        }
+
+        logger.debug(
+            `Figure out conn status for ${sourceName}, is video muted: ${
+                isVideoMuted} video track frozen: ${
+                isVideoTrackFrozen} p2p mode: ${
+                inP2PMode} is in forwarded sources: ${
+                isInForwardedSources} currentStatus => newStatus: ${
+                this.track.getConnectionStatus()} => ${newState}`);
+
+        const oldConnectionStatus = this.connectionStatusMap || {};
+
+        // Send an analytics event (guard on either the p2p flag or the connection status has changed
+        // since the last time this code block run).
+        if (!('p2p' in oldConnectionStatus)
+            || !('connectionStatus' in oldConnectionStatus)
+            || oldConnectionStatus.p2p !== inP2PMode
+            || oldConnectionStatus.connectionStatus !== newState) {
+
+            const nowMs = Date.now();
+
+            this.maybeSendTrackConnectionStatusEvent(nowMs);
+
+            this.connectionStatusMap = {
+                ...oldConnectionStatus,
+                connectionStatus: newState,
+                p2p: inP2PMode,
+                startedMs: nowMs
+            };
+
+            // sometimes (always?) we're late to hook the TRACK_VIDEOTYPE_CHANGED event and the
+            // video type is not in oldConnectionStatus.
+            if (!('videoType' in this.connectionStatusMap)) {
+                this.connectionStatusMap.videoType = this.track.getVideoType();
+            }
+        }
+        this._changeConnectionStatus(newState);
+    }
+
+    /**
+     * Computes the duration of the current connection status for the track (i.e. 15 seconds
+     * in the INTERRUPTED state) and sends a track connection status event.
+     * @param {Number} nowMs - The current time (in millis).
+     * @returns {void}
+     */
+    maybeSendTrackConnectionStatusEvent(nowMs) {
+        const trackConnectionStatus = this.connectionStatusMap;
+
+        if (trackConnectionStatus
+            && 'startedMs' in trackConnectionStatus
+            && 'videoType' in trackConnectionStatus
+            && 'connectionStatus' in trackConnectionStatus
+            && 'p2p' in trackConnectionStatus) {
+            trackConnectionStatus.value = nowMs - trackConnectionStatus.startedMs;
+            Statistics.sendAnalytics(
+                createTrackConnectionStatusEvent(trackConnectionStatus));
+        }
+    }
+
+    /**
+     * On change in forwarded sources set check all leaving and entering track to
+     * change their corresponding statuses.
+     *
+     * @param {Array<string>} leavingForwardedSources - The array of sourceName leaving forwarded sources.
+     * @param {Array<string>} enteringForwardedSources - The array of sourceName entering forwarded sources.
+     * @private
+     */
+    _onForwardedSourcesChanged(leavingForwardedSources = [], enteringForwardedSources = []) {
```

Comment:
I don't think this will work as I expect it to. I'll need to move this logic somewhere else, perhaps `JitsiConferenceEventManager`, and have it iterate through all the leaving and entering forwarded sources.

Reply:
Ah because the timestamps will not be accurate since this instance may disappear. Lets try to think about it... Maybe we need to store the timestamps somewhere (the place which emits the forwarded sources changed event) and then expose that via a getter? Or better have a static part/helper class of `TrackStreamingStatus` which would keep track of the timestamps (mostly because they are not used by any other logic) ?

Reply:
I was looking into using a static method/helper class and it was starting to get a bit messy since it would still need to call `figureOutStreamingStatus` for each instances that is leaving/entering or when `!browser.supportsVideoMuteOnConnInterrupted()`. If our only concern is having inaccurate timestamps and performance isn't an issue, can we just emit the timestamp with the forwarded sources?

Reply:
We can actually make it so that the `JitsiRemoteTrack` holds the timestamps when it entered/left the forwarded sources and this can happen any time the `RTC` module emits the forwarded sources event?